Improve utility functions

akukanov opened the pull request:
The patch refactors the SIMD utility functions for load, store, gather, and scatter:

* Removes `enable_if` SFINAE, instead using specific parameter types
* Aligns the API semantics (parameter order, types, and names)
* For `gather`, also fixes the result type to differ from offsets, and moves items-to-bytes conversion to the caller (consistent to `scatter`)

Also it removes unneeded `inline` specifiers and adds missed standard headers.

akukanov commented on include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:
```diff
@@ -3,79 +3,73 @@
+gather(T* input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets, ::std::uint32_t base_offset)
 {
-    return sycl::ext::intel::esimd::gather(input + base_offset, offsets*static_cast<uint32_t>(sizeof(T)));
+    return sycl::ext::intel::esimd::gather(input + base_offset, offsets);
```
Computing the base address is incorrect if the offset is in bytes (which it is supposed to be).

Perhaps it makes sense to move to the element-based offset in all functions, converting to bytes when required by ESIMD API.

akukanov replied:
Changed to use element-based offsets in the utilities. Note that the accessor variants of copy_to/from now also convert it to bytes - I believe this matches the ESIMD API, essentially fixing a bug.

SergeyKopienko reviewed (changes requested):
copy_from(const T* input, ::std::uint32_t base_offset, sycl::ext::intel::esimd::simd<T, N>& values)

SergeyKopienko commented:
@akukanov unfortunately there are many compile errors now

akukanov commented:
> @akukanov unfortunately there are many compile errors now

Fixed.

> copy_from(const T* input, ::std::uint32_t base_offset, sycl::ext::intel::esimd::simd<T, N>& values)
> gather(const T* input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets, ::std::uint32_t base_offset)

Fixed as well.

SergeyKopienko commented on include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:
```diff
@@ -3,79 +3,73 @@
+        sycl::ext::intel::esimd::simd<T, N> values, sycl::ext::intel::esimd::simd_mask<N> mask = 1)
 {
-    return sycl::ext::intel::esimd::scatter(input, offsets, vals, mask);
+    return sycl::ext::intel::esimd::scatter(output, offsets, values, mask);
```
```return``` not required here

dmitriy-sobolev commented on include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:
```diff
@@ -3,79 +3,84 @@
+    //static_assert(::std::is_same<decltype(offsets), decltype(offsets * sizeof(T))>::value, "'offsets' and 'offsets * sizeof(T)' are not the same");
+    //static_assert(::std::is_same<decltype(base_offset), decltype(base_offset * sizeof(T))>::value, "'base_offset' and 'base_offset * sizeof(T)' are not the same");
+
+    return sycl::ext::intel::esimd::gather<T>(input, offsets /** sizeof(T)*/, base_offset /** sizeof(T)*/);
```
The comment is the same as in https://github.com/oneapi-src/oneDPL/pull/871/files#r1156252968, but the imitated API is [that](https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory.html#gab8e33bce282903e5c4e74b009ed0d0c9). Both offsets should be multiplied by `sizeof(T)` because `gather` expect offsets expressed in bytes. Static cast is needed to explicitly pass arguments with types which ESIMD expects.
```suggestion
    return sycl::ext::intel::esimd::gather<T>(input, offsets * static_cast<::std::uint32_t>(sizeof(T)), base_offset * static_cast<::std::uint32_t>(sizeof(T)));
```

SergeyKopienko replied:
Fixed

SergeyKopienko commented on include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:
```diff
@@ -3,79 +3,84 @@
+    //decltype(offsets)::dummy;
+    //decltype(sizeof(T))::dummy;
+    //decltype(offsets * sizeof(T))::dummy;
+    return sycl::ext::intel::esimd::gather(input + base_offset, offsets/* * sizeof(T)*/);
```
Fixed

SergeyKopienko reviewed (approved):
Now it's compiled and LGTM

akukanov commented:
Thank you both for the help with fixing the issues. The changes were fine, but I still wanted to completely avoid SFINAE to only allow the API to be used with accessors and pointers, and also I simplified the way to deal with `sizeof` causing SIMD type change.
